From 77a1da74a8da7447d0784494cc1ef06ec61467a7 Mon Sep 17 00:00:00 2001
From: Alison Wang <alison.wang@nxp.com>
Date: Wed, 15 Aug 2018 14:11:11 +0800
Subject: [PATCH 262/706] drm: hdp: Add reset operations for eDP PHY

This patch adds reset operations for eDP PHY.

Signed-off-by: Alison Wang <alison.wang@nxp.com>
(cherry picked from commit 8b4c585f5baa64e01cd52338343d7d56d02f7dff)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-dp.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/drivers/gpu/drm/imx/hdp/imx-dp.c b/drivers/gpu/drm/imx/hdp/imx-dp.c
index 311dd5bbcd2a..2db6216f1202 100644
--- a/drivers/gpu/drm/imx/hdp/imx-dp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-dp.c
@@ -6,11 +6,16 @@
  */ 
 
 #include <linux/clk.h>
+#include <linux/io.h>
+#include <linux/of.h>
+#include <linux/of_address.h>
 #include "mhdp_firmware.h"
 #include "imx-hdp.h"
 #include "imx-hdmi.h"
 #include "imx-dp.h"
 
+#define EDP_PHY_RESET	0x230
+
 void dp_fw_load(state_struct *state)
 {
 	CDN_API_LoadFirmware(state,
@@ -45,9 +50,22 @@ void dp_fw_init(state_struct *state, u32 core_rate)
 		ADDR_SOURCD_PHY + (LANES_CONFIG << 2), 0x0040001b);
 }
 
+static const struct of_device_id scfg_device_ids[] = {
+	{ .compatible = "fsl,ls1028a-scfg", },
+	{}
+};
+
 void dp_phy_init(state_struct *state, int num_lanes, int max_link_rate, int tmp)
 {
 	int ret;
+	struct device_node *scfg_node;
+	void __iomem *scfg_base = NULL;
+
+	scfg_node = of_find_matching_node(NULL, scfg_device_ids);
+	if (scfg_node)
+		scfg_base = of_iomap(scfg_node, 0);
+
+	iowrite32(0, scfg_base + EDP_PHY_RESET);
 
 	/* PHY initialization while phy reset pin is active */
 	AFE_init(state, num_lanes, (ENUM_AFE_LINK_RATE)max_link_rate);
@@ -57,6 +75,9 @@ void dp_phy_init(state_struct *state, int num_lanes, int max_link_rate, int tmp)
 	hdp_phy_reset(1);
 #endif
 
+	iowrite32(0x1, scfg_base + EDP_PHY_RESET);
+	iounmap(scfg_base);
+
 	/* PHY power set */
 	AFE_power(state, num_lanes, (ENUM_AFE_LINK_RATE)max_link_rate);
 
-- 
2.17.1

